Componentize UI: introduce Note + ChecklistItem components, and refactor boards page

tsahil01 opened the pull request:
fix #128 


**Type Standardization and Refactoring:**

* Replaced inline interface/type definitions for `User`, `Board`, `ChecklistItem`, `Note`, and related types in multiple files (`app/settings/layout.tsx`, `app/settings/page.tsx`, `app/settings/organization/page.tsx`, `app/dashboard/page.tsx`, `app/public/boards/[id]/page.tsx`, `app/api/boards/[id]/notes/[noteId]/route.ts`) with imports from `@/lib/types`. This ensures consistent type usage throughout the app. [[1]](diffhunk://#diff-8bff06dd32335e275d4fcb561da73c0b4ecba21ddc5cf10c850273a9d582d0e4L5-R9) [[2]](diffhunk://#diff-8bff06dd32335e275d4fcb561da73c0b4ecba21ddc5cf10c850273a9d582d0e4L162-R148) [[3]](diffhunk://#diff-e441de59e2b06b106855d24f892efc380ce6620dfe254590fd5debb910467ce6L11-R11) [[4]](diffhunk://#diff-54f499228aff9c44898e94b97536dab6fb46b9919a7c3e0befe33fd61b153d37L32-R32) [[5]](diffhunk://#diff-54f499228aff9c44898e94b97536dab6fb46b9919a7c3e0befe33fd61b153d37L73-R57) [[6]](diffhunk://#diff-54f499228aff9c44898e94b97536dab6fb46b9919a7c3e0befe33fd61b153d37L585-R569) [[7]](diffhunk://#diff-e16c1c1133d5d72341ada507f3e06fa37dd79de574403f158159ff934f9b5013L37-R41) [app/public/boards/[id]/page.tsxL10-R10](diffhunk://#diff-f027368de174318180834854c49d03296e08d413eaf700c2dcbb4750a6c598ccL10-R10), [app/api/boards/[id]/notes/[noteId]/route.tsL5-R5](diffhunk://#diff-2954d4fd354cb92e0bcf0fe9da689b7ddbf4766acc51885b7f6d6a30cdc20205L5-R5))

**Componentization and Reusability:**

* Added new reusable client-side components: `ChecklistItem` (`components/checklist-item.tsx`) and `Note` (`components/note.tsx`), encapsulating UI and logic for checklist items and notes, respectively. These components support editing, deletion, and other interactions, improving modularity and reusability. [[1]](diffhunk://#diff-f7af254e695b6beaa8ea1878b56a86a92b28a989795fa70f703bccb753a14b9dR1-R125) [[2]](diffhunk://#diff-58a3ef678b0dd022587036ba02cf5387acb26d2599eed844b494bff20f71b8bfR1-R278)
* Updated `components/index.ts` to export the new components and relevant types for convenient reuse across the application.

**UI Improvements:**

* Adjusted icon imports and usages (e.g., `User` to `UserIcon`) for naming clarity and to resolve naming conflicts in components. [[1]](diffhunk://#diff-8bff06dd32335e275d4fcb561da73c0b4ecba21ddc5cf10c850273a9d582d0e4L5-R9) [[2]](diffhunk://#diff-8bff06dd32335e275d4fcb561da73c0b4ecba21ddc5cf10c850273a9d582d0e4L162-R148)

These changes collectively make the codebase more maintainable, reduce duplication, and streamline the development of features involving notes and checklist items.

**New interactive checklist note system:**

* Added `ChecklistItem` component to render and manage individual checklist items with editing, toggling, splitting, and deleting capabilities. (`components/checklist-item.tsx`)
* Added `Note` component to display a note with user info, editable content, and an interactive checklist, including permissions logic for editing and deleting. (`components/note.tsx`)
* Added `NotesDemo` component to demonstrate interactive checklist notes with sample data, supporting add/edit/delete for notes and checklist items, and animated transitions. (`components/notes-demo.tsx`)

**Type and demo data refactoring:**

* Updated sticky note demo components to use shared `DemoNote` and `DemoTask` types from `@/lib/types` instead of local definitions, improving type consistency and maintainability. (`components/sticky-note-card.tsx`, `components/sticky-notes-demo.tsx`) [[1]](diffhunk://#diff-47b82e4af6fabdbb8bb1ce84b9f30f66eb1177f7e4165902ecd9b6b827ec96f9R13-R17) [[2]](diffhunk://#diff-400471a9138e1c457c219bf5aac2cdf17ef21a1622ab11b31ec28cdec12aeb34R7-R11) [[3]](diffhunk://#diff-400471a9138e1c457c219bf5aac2cdf17ef21a1622ab11b31ec28cdec12aeb34L274-R277) [[4]](diffhunk://#diff-400471a9138e1c457c219bf5aac2cdf17ef21a1622ab11b31ec28cdec12aeb34L287-R288)

**Component and type exports:**

* Centralized exports for all reusable components and related types in a new `components/index.ts` file for easier imports and improved code organization. (`components/index.ts`)

**Tests:**
<img width="417" height="173" alt="image" src="https://github.com/user-attachments/assets/98e47594-dc91-4c6d-8137-b6355286f30e" />


All tests passed.

tsahil01 commented:
Resolved build errors.

tsahil01 commented:
@slavingia Can we merge this?

tsahil01 commented:
@slavingia all 32 playwright tests are passed.

<img width="1092" height="946" alt="image" src="https://github.com/user-attachments/assets/fdb4e90b-b2ea-4fe2-b79f-19f286e34c9e" />
<img width="1427" height="1041" alt="image" src="https://github.com/user-attachments/assets/098db0e4-5dae-4176-a87b-c366dbec08c7" />

slavingia commented on app/dashboard/page.tsx:
```diff
@@ -34,32 +34,11 @@ import {
-  } | null;
-}
+import type { DashboardBoard, User, Board } from "@/lib/types";
+// keep Board for editingBoard state
```
?

tsahil01 replied:
it's unnecessary. removing it now.

slavingia commented:
Can this be broken up into any sub PRs, some of which may be easier to merge? Don't think types should get their own file, better to be exported from the respective component itself imo if needed in other places.

Please also self-review your code.

tsahil01 commented:
@slavingia Done! I've removed the centralized types file and moved everything back to components like you suggested. 
Deleted lib/types.ts and put the main types (Note, ChecklistItem, etc.) in components/note.tsx instead. 
Dashboard and settings pages now define their own extended types locally, and I updated all the imports. 
TypeScript compiles fine and all tests pass. 

Tests: (32/32 passed)
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/9c44d087-bf70-40b1-af18-2ee72f813060" />

slavingia commented on components/checklist-item.tsx:
```diff
@@ -0,0 +1,125 @@
+        </span>
+      )}
+
+      {/* Delete Button */}
```
imo these kinds of comments aren't helpful and can be removed

slavingia commented on components/sticky-notes-demo.tsx:
```diff
@@ -311,7 +443,19 @@ export function StickyNotesDemo() {
+                  onEditChecklistItem={handleEditChecklistItem}
+                  onDeleteChecklistItem={handleDeleteChecklistItem}
+                  onToggleAllChecklistItems={handleToggleAllChecklistItems}
+                  className={`${note.color} bg-white dark:bg-zinc-900`}
+                  style={{ padding: "16px" }}
```
Can we move these to the component (using tailwind instead of explicit 16px) so that it matches the app and vice versa perfectly?

slavingia reviewed (commented):
Overall looks great, minor comments!

tsahil01 commented:
@slavingia Fixed all the issues you mentioned!
Ready for merge?

sm17p commented on app/dashboard/page.tsx:
```diff
@@ -34,32 +34,19 @@ import {
 
 export default function Dashboard() {
-  const [boards, setBoards] = useState<Board[]>([]);
+  const [boards, setBoards] = useState<DashboardBoard[]>([]);
```
Damn! This should've been called `GumBoard` haha
